core/src/main/java/hudson/FilePath.java

```diff
@@ -563,39 +584,64 @@ public Void invoke(File dir, VirtualChannel channel) throws IOException, Interru
      * @param compression
      *      Compression mode of this tar file.
      * @since 1.292
+     *
+     * @see FilePath#untar(FilePath, TarCompression, Consumer)
      * @see #untarFrom(InputStream, TarCompression)
      */
     public void untar(final FilePath target, final TarCompression compression) throws IOException, InterruptedException {
+        untar(target, compression, (Consumer<List<String>> & Serializable)(a) -> {});
+    }
+
+    /**
+     * When this {@link FilePath} represents a tar file, extracts that tar file.
+     *
+     * @param target
+     *      Target directory to expand files to. All the necessary directories will be created.
+     * @param compression
+     *      Compression mode of this tar file.
+     * @param consumer
+     *      Consumer for processing the untar files.
+     *
+     * @since 2.267
+     * @see #untarFrom(InputStream, TarCompression)
+     */
+    public void untar(final FilePath target, final TarCompression compression, Consumer<List<String>> consumer) throws IOException, InterruptedException {
         // TODO: post release, re-unite two branches by introducing FileStreamCallable that resolves InputStream
         if (this.channel!=target.channel) {// local -> remote or remote->local
             final RemoteInputStream in = new RemoteInputStream(read(), Flag.GREEDY);
-            target.act(new UntarRemote(compression, in));
+            target.act(new UntarRemote(compression, in, consumer));
         } else {// local -> local or remote->remote
-            target.act(new UntarLocal(compression));
+            target.act(new UntarLocal(compression, consumer));
         }
     }
     private class UntarRemote extends SecureFileCallable<Void> {
         private final TarCompression compression;
         private final RemoteInputStream in;
-        UntarRemote(TarCompression compression, RemoteInputStream in) {
+        private final Consumer<List<String>> consumer;
+
+        UntarRemote(TarCompression compression, RemoteInputStream in, Consumer<List<String>> consumer) {
             this.compression = compression;
             this.in = in;
+            this.consumer = consumer;
         }
         @Override
         public Void invoke(File dir, VirtualChannel channel) throws IOException, InterruptedException {
-            readFromTar(FilePath.this.getName(), dir, compression.extract(in));
+            consumer.accept(readFromTar(FilePath.this.getName(), dir, compression.extract(in)));
```

Comment:
if you want logging here - then I would assume that you want logging as it happens not after the fact (if you want after the fact you could as easily `ls` or the equivallent?